First steps with Jumpstarter (#1)
rhpds/showroom-lb2255-jumpstarter: content/modules/ROOT/pages/module-03.adoc

= First steps with Jumpstarter

> Jumpstarter is already pre-installed in the developer workspace, you can check for it with

Reviewer: ```suggestion
Jumpstarter is already pre-installed and pre-configured in the developer workspace, so it should already be available in the console.

To check that Jumpstarter is installed, run the following from the editor terminal:
```

[source,bash]
----
jmp version
----

> Jumpstarter interacts with hardware interfaces through "drivers", drivers are written in python to make use of the vast ecosystem of existing python packages. While you can always extend Jumpstarter by writing your own drivers, it comes with a collection of general purpose drivers to help you get started.

Reviewer: ```suggestion
Jumpstarter interacts with hardware interfaces through "drivers", drivers are written in Python to make use of the vast ecosystem of existing Python packages.
While you can always extend Jumpstarter by writing your own drivers, it comes with a collection of general-purpose drivers to help you get started.

To list the available drivers, run the following command:
```

[source,bash]
----
jmp driver list
----

> TIP: Among the drivers you would notice a few with "mock" in their names, while these mock drivers do not provide access to the hardware, they are useful in showcasing the capability of Jumpstarter

Reviewer: ```suggestion
[TIP]
====
Some driver packages provide mock drivers, usually with a `Mock` prefix. While mock drivers to not provide access to physical or virtual hardware, they are useful for demonstrating the capabilities of Jumpstarter.
```

[#config]
== Configuring Jumpstarter for local development

> Create a new file in the workspace named `first-steps.yaml`, this would be an "exporter config file", in which we will configure a list of drivers representing the hardware interfaces we want to "export" to the users (in this case, ourselves).

Reviewer: ```suggestion
Create a new file in the workspace named `first-steps.yaml`, this will be our exporter config file. We can use this file to configure a list of drivers representing the hardware interfaces we want to "export" to the Jumpstarter clients.
```

> Create a new file in the workspace named `first-steps.yaml`, this would be an "exporter config file". Within which we would configure a list of drivers representing the hardware interfaces we would like to "export" to the users (that happens to be ourselves in this case).

Reviewer 2: ```suggestion
Create a new file in the workspace named `first-steps.yaml`, this would be an "exporter config file", in which we will configure a list of drivers representing the hardware interfaces we want to "export" to the users (in this case, ourselves).
```

.first-steps.yaml
[source,yaml]
----
apiVersion: jumpstarter.dev/v1alpha1
kind: ExporterConfig
metadata:
  namespace: default
  name: first-steps
endpoint: ""
token: ""
export:
  power:
    # provides power management: power on/off/cycle
    type: jumpstarter_driver_power.driver.MockPower
  console:
    # provides serial console access
    type: jumpstarter_driver_pyserial.driver.PySerial
    config:
      url: "loop://" # a mock serial port that echos everything
  flasher:
    # provides storage access: like flashing a disk image into a sdcard
    type: jumpstarter_driver_opendal.driver.MockFlasher
----

NOTE: endpoint and token are intentionally left empty here, since they are only useful for remote access workflows.

> Now we can enter the exporter shell, within which a transient Jumpstarter exporter will be started and made available for us to interact with.

Reviewer: ```suggestion
To test out this config file, we can create a transient Jumpstarter exporter shell. This allows us to test an exporter config and drivers locally before deploying them into a testing environment.
```

> Now we can enter the exporter shell, within which a transient Jumpstarter exporter would be started and made available for us to interact with.

Reviewer 2: ```suggestion
Now we can enter the exporter shell, within which a transient Jumpstarter exporter will be started and made available for us to interact with.
```

[source,bash]
----
jmp shell --exporter-config ./first-steps.yaml
----

TIP: Notice that you shell prompt have been changed to `jumpstarter ⚡local ➤`, this is to remind you that you are inside a local Jumpstarter session.

[#interact]
== Interacting with a local Jumpstarter session with CLI

> A `j` command is available inside `jmp shell`, it is dynamically constructed reflecting the drivers configured on the exporter side. Not all drivers have corresponding CLI interfaces, but it is available on most builtin drivers.

Reviewer: ```suggestion
A `j` command is available inside `jmp shell`, it is dynamically constructed reflecting the drivers configured on the exporter side. Not all drivers have corresponding CLI interfaces, but it is available on most built-in drivers.
```

> A `j` command is available inside `jmp shell`, it is dynamically constructed reflecting the drivers configured on the exporter side. Not all dirvers have corresponding CLI interfaces, but it is available on most builtin drivers.

Reviewer 2: ```suggestion
A `j` command is available inside `jmp shell`, it is dynamically constructed reflecting the drivers configured on the exporter side. Not all drivers have corresponding CLI interfaces, but it is available on most builtin drivers.
```

[source,bash]
----
j
----

[NOTE]
====

> With our particular configuration in `first-steps.yaml`, you should see something like

Reviewer: ```suggestion
With the exporter config defined in `first-steps.yaml`, we should see an output like this:
```

....
Usage: j [OPTIONS] COMMAND [ARGS]...

  Generic composite device

Options:
  --help  Show this message and exit.

Commands:
  console  Serial port client
  flasher  Generic flasher interface
  power    Generic power
....
====

Try turning the power of the device under test on and off

[source,bash]
----
j power on
j power off
----

> TIP: Nothing actually happens because we are using a mock power driver, but have we configured it with a real power driver, e.g. a USB power relay, you could see the device under test being turned on and off.

Reviewer 2: ```suggestion
TIP: Nothing actually happens because we are using a mock power driver, but have we configured it with a real power driver, e.g. a USB power relay, you would see the device under test being turned on and off.
```

> In this secion, we've learned how to configure Jumpstarter for local developement and how to interact with the device under test via the `j` command. If you still have time, explore the other subcommands (`j flasher`/`j console`)

Reviewer 2: ```suggestion
In this secion, we've learned how to configure Jumpstarter for local development and how to interact with the device under test via the `j` command. If you still have time, explore the other subcommands (`j flasher`/`j console`)
```
